src/Std/Sat/AIG/Basic.lean

```diff
@@ -0,0 +1,506 @@
+/-
+Copyright (c) 2024 Lean FRO, LLC. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Henrik Böving
+-/
+import Std.Data.HashMap
+import Std.Data.HashSet
+
+namespace Std
+namespace Sat
+
+/-!
+This module contains the basic definitions for an AIG (And Inverter Graph) in the style of AIGNET,
+as described in https://arxiv.org/pdf/1304.7861.pdf section 3. It consists of an AIG definition,
+a description of its semantics and basic operations to construct nodes in the AIG.
+-/
+
+variable {α : Type} [Hashable α] [DecidableEq α]
+
+namespace AIG
+
+/--
+A circuit node declaration. These are not recursive but instead contain indices into an `AIG`, with inputs indexed by `α`.
+-/
+inductive Decl (α : Type) where
+  /--
+  A node with a constant output value.
+  -/
+  | const (b : Bool)
+  /--
+  An input node to the circuit.
+  -/
+  | atom (idx : α)
+  /--
+  An AIG gate with configurable input nodes and polarity. `l` and `r` are the
+  input node indices while `linv` and `rinv` say whether there is an inverter on
+  the left or right input.
+  -/
+  | gate (l r : Nat) (linv rinv : Bool)
+  deriving Hashable, Repr, DecidableEq, Inhabited
+
+
+/--
+A `Cache` is valid with respect to a list of declarations iff:
+Whenever `cache.find? decl` returns an index into `xs : Array Decl`, `xs[index] = decl`.
+This predicate limits a `HashMap Decl Nat` to this behavior.
```

Comment:
```suggestion
`Cache.WF xs` is a predicate asserting that a `cache : HashMap (Decl α) Nat`
is a valid lookup cache for `xs : Array (Decl α)`, that is,
whenever `cache.find? decl` returns an index into `xs : Array Decl`, `xs[index] = decl`.
```

(My concern here is that your original docstring starts by talking about `Cache`, which hasn't otherwise been mentioned yet.)

Reply:
Maybe explicitly say that it doesn't need to be complete? Something like
"It need not be complete, in the sense that `cache.find? decl = none` is allowed even when `decl \mem xs`."